src/erlcloud_aws.erl

```diff
@@ -901,52 +901,72 @@ profiles_parse( Content ) ->
+profiles_resolve( Name, Profiles, Role, ExternalId, Options ) ->
     case proplists:get_value( Name, Profiles ) of
         undefined ->
             error_msg( "profile ~s does not exist", [Name] );
         Keys ->
-            profiles_recurse( Keys, Profiles, Role, Options )
+            profiles_recurse( Keys, Profiles, Role, ExternalId, Options )
     end.
 
-profiles_recurse( Keys, Profiles, Role, Options ) ->
+profiles_recurse( Keys, Profiles, Role, ExternalId, Options ) ->
     case profiles_credentials( Keys ) of
         {ok, Credential} ->
-            profiles_assume( Credential, Role, Options );
-        {cont, ProfileName, BinRole} ->
-            profiles_resolve( ProfileName, Profiles, BinRole, Options )
+            profiles_assume( Credential, Role, ExternalId, Options );
+        {cont, ProfileName, NextRole, NextExternalId} ->
+            case profiles_resolve( ProfileName, Profiles,
+                                   NextRole, NextExternalId, Options ) of
+                {ok, Config} ->
+                    profiles_assume( Config, Role, ExternalId, Options );
+                Otherwise -> Otherwise
+            end
     end.
 
 profiles_credentials( Keys ) ->
-    Names = [aws_access_key_id, aws_secret_access_key, aws_security_token, source_profile],
-    BinRole = proplists:get_value( role_arn, Keys ),
+    Names = [aws_access_key_id, aws_secret_access_key, aws_security_token,
+             source_profile],
     case [proplists:get_value( K, Keys ) || K <- Names] of
         [Id, Secret, undefined, undefined] when Id =/= undefined, Secret =/= undefined ->
             {ok, {binary_to_list(Id), binary_to_list(Secret)}};
         [Id, Secret, Token, undefined] when Id =/= undefined, Secret =/= undefined, Token =/= undefined ->
             {ok, {binary_to_list(Id), binary_to_list(Secret), binary_to_list(Token)}};
-        [undefined, undefined, undefined, BinProfile] when BinProfile =/= undefined ->
-            {cont, BinProfile, BinRole}
+        [undefined, undefined, undefined, SourceProfile] ->
+            profiles_credentials( Keys, SourceProfile )
     end.
 
-profiles_assume( Credential, undefined, _Options ) ->
+profiles_credentials( Keys, SourceProfile ) ->
+    Names = [role_arn, external_id],
+    [RoleArn, ExternalId] = [proplists:get_value( K, Keys ) || K <- Names],
+    {cont, SourceProfile, RoleArn, ExternalId}.
+
+profiles_assume( Credential, undefined, __ExternalId, _Options ) ->
     Config = config_credential(Credential, #aws_config{}),
     {ok, Config};
-profiles_assume( Credential, Role,
-                 #profile_options{ session_name = Name, external_id = ExtId,
+profiles_assume( Credential, Role, ExternalId,
+                 #profile_options{ session_name = Name,
+                                   external_id = DefaultExternalId,
                                    session_secs = Duration } ) ->
+    ExtId = if ExternalId =/= undefined -> ExternalId;
```

Comment:
doing the second clause as the opposite of the first is a stylistic thing, rather than just saying `true` for the second clause.  I feel that being explicit does not allow unexpected variations to pass through, as would happen by just ending with `true`.